### PR TITLE
[#5982] feat (gvfs-fuse): Implement Gravitino fileset file system

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

Implement an Gravitino fileset file system, Support mount fileset to local directory

### Why are the changes needed?

Fix: #5982

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT and IT

### CODE REVIEW

File: clients/filesystem-fuse/src/gravitino_client.rs [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+use crate::config::GravitinoConfig;
+use crate::error::{ErrorCode, GvfsError};
+use reqwest::Client;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use urlencoding::encode;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct Fileset {
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gravitino_client.rs:28]: could you move fileset related struct to sepreate file?

File: clients/filesystem-fuse/src/gravitino_client.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+pub(crate) struct GravitinoClient {
+    gravitino_uri: String,
+    metalake: String,
+
+    client: Client,
+}
+
+impl GravitinoClient {
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gravitino_client.rs:58]: The client is binded to fileset,   Could you plan to add some abstractions like `loadCatalog` like java client for more extensiable.

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        _context: &FileSystemContext,
+    ) -> Self {
+        Self {
+            fs: fs,
+            client: client,
+            fileset_location: location.into(),
+        }
+    }
+
+    fn map_fileset_path_to_raw_path(&self, path: &Path) -> PathBuf {
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gvfs_fileset_fs.rs:53]: the method name seems too long, how about `to_raw_path` and `to_fileset_path`?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+use async_trait::async_trait;
+use fuse3::Errno;
+use std::path::{Path, PathBuf};
+
+pub(crate) struct GravitinoFileSystemConfig {}
+
+/// GravitinoFileSystem is a filesystem that is associated with a fileset in Gravitino.
+/// It mapping the fileset path to the original data storage path. and delegate the operation
+/// to the inner filesystem like S3 GCS, JuiceFS.
+pub(crate) struct GvfsFilesetFs {
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gvfs_fileset_fs.rs:32]: How about  rename `GvfsFilesetFs` to `FilesetFilesystem`?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    fs: Box<dyn PathFileSystem>,
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gvfs_fileset_fs.rs:33]: `fs` is too gneneral, use a better name?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 4 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs:32]: If that, `GravitinoFilesetFilesystem` is better. another filesystem associated with catalog/schema named `GravitinoFilesystem`, is Ok?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 5 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs:33]: Is `inner_fs` ok?

File: clients/filesystem-fuse/src/gravitino_client.rs (hunk 2 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gravitino_client.rs:58]: Currently, I think it's unnecessary to make it so complex. Just providing the required `fileset` API here is sufficient.

File: clients/filesystem-fuse/src/gravitino_client.rs (hunk 1 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gravitino_client.rs:28]: I haven't figured out how to organize these structures yet; they will only be used by the gravitino client. Do you have any ideas?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 4 above)
Comment (FANNG1) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs:32]: I prefer `FilesetFilesystem`, but I'm ok to `GravitinoFilesetFilesystem` if you have strong option.

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 5 above)
Comment (FANNG1) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs:33]: something like `physical_fs`?  the underlying fs are doing real physical operations

File: clients/filesystem-fuse/src/gravitino_client.rs (hunk 2 above)
Comment (FANNG1) [reply clients/filesystem-fuse/src/gravitino_client.rs:58]: @jerryshao , WDYT?

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 5 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs]: fix

File: clients/filesystem-fuse/src/gvfs_fileset_fs.rs (hunk 4 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gvfs_fileset_fs.rs]: fix

File: clients/filesystem-fuse/src/gvfs_fuse.rs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+use once_cell::sync::Lazy;
+use std::path::Path;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+const FILESET_PREFIX: &str = "gvfs://fileset/";
+
+static SERVER: Lazy<Mutex<Option<Arc<FuseServer>>>> = Lazy::new(|| Mutex::new(None));
+
+pub(crate) enum CreateFsResult {
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/gvfs_fuse.rs:39]: `CreateFsResult` is a little odd, please provide a better name

File: clients/filesystem-fuse/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+use crate::utils::GvfsResult;
+use config::{builder, Config};
+use log::{error, info, warn};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fs;
+
+pub(crate) const CONF_FUSE_FILE_MASK: ConfigEntity<u32> = ConfigEntity::new(
+    FuseConfig::MODULE_NAME,
+    "file_mask",
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/config.rs:29]: No need to add a prefix. It is below `FuseConfig::MODULE_NAME`.

File: clients/filesystem-fuse/src/config.rs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+pub(crate) const CONF_GRAVITINO_URL: ConfigEntity<&'static str> = ConfigEntity::new(
+    GravitinoConfig::MODULE_NAME,
+    "gravitino_url",
+    "The URL of the Gravitino server",
+    "http://localhost:8090",
+);
+
+pub(crate) const CONF_GRAVITINO_METALAKE: ConfigEntity<&'static str> = ConfigEntity::new(
+    GravitinoConfig::MODULE_NAME,
+    "metalake",
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/config.rs:71]: No need to add a prefix. It is below `GravitinoConfig::MODULE_NAME`.

File: clients/filesystem-fuse/src/config.rs [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+pub(crate) const CONF_FILESYSTEM_BLOCK_SIZE: ConfigEntity<u32> = ConfigEntity::new(
+    FilesystemConfig::MODULE_NAME,
+    "block_size",
+    "The block size of the gvfs fuse filesystem",
+    4096,
+);
+
+pub(crate) const CONF_GRAVITINO_URL: ConfigEntity<&'static str> = ConfigEntity::new(
+    GravitinoConfig::MODULE_NAME,
+    "gravitino_url",
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/config.rs:64]: No need to add a prefix. It is below `GravitinoConfig::MODULE_NAME`.

File: clients/filesystem-fuse/src/gvfs_fuse.rs (hunk 6 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/gvfs_fuse.rs:39]: `CreateFileSystemResult` is better?

File: clients/filesystem-fuse/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply clients/filesystem-fuse/src/config.rs:71]: it does not consist of `gravitino_uri`,

[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/config.rs:71]: change to `uri`

**Review by FANNG1 — APPROVED** [review]

**Pull request merged**